Subscriptions: 27: Update navigation to use Push Views

afterxleep opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.

⚠️ If you're an external contributor, please file an issue first before working on a PR, as we can't guarantee that we will accept your changes if they haven't been discussed ahead of time. Thanks!
-->

Task/Issue URL: https://app.asana.com/0/414235014887631/1206942263287858/f

**Description**:
- Updates Subscription navigation to use "Push" transitions instead of sheets
- Adds `.onFirstAppear()` modifier to view as SwiftUI's `.onAppear()` might be called multiple times
- Cache subscription in memory to prevent UI Glitches
- Other minor fixes and renames

**Steps to test this PR**:
1. Subscribe Via App Store
2. Once in the welcome page, navigate through the different features (VPN, ITP, etc)
3. Add Email to subscription
4. Remove Subscription from device
5. Restore subscription via email
6. Once in the welcome page, navigate through the different features (VPN, ITP, etc)


**Demo**:
![RocketSim_Recording_iPhone_(17 2)_6 1_2024-03-29_12 04 56](https://github.com/duckduckgo/iOS/assets/1156669/08bd5256-1f16-40cc-970b-2c7970717b19)


<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable.
-->
---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

afterxleep committed:
Reset State before setting up subscription environment

afterxleep committed:
Inject UserScript down the stack as a single instance

afterxleep committed:
Subscription Flow Base View

afterxleep committed:
Merge branch 'main' into daniel/subscriptions/25.push.views

afterxleep committed:
Onboarding Buttons Working & Add Email

afterxleep committed:
Personal Information Removal Stack

afterxleep committed:
ITP Management from Webview

afterxleep committed:
Restore View + App Store Restore

afterxleep committed:
Navigation Updates for Email

afterxleep committed:
Merge branch 'main' into daniel/subscriptions/27.push.views

# Conflicts:
#	DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved
#	DuckDuckGo/SettingsSubscriptionView.swift
#	DuckDuckGo/Subscription/ViewModel/SubscriptionFlowViewModel.swift
#	DuckDuckGo/Subscription/ViewModel/SubscriptionRestoreViewModel.swift
#	DuckDuckGo/Subscription/Views/SubscriptionFlowView.swift
#	DuckDuckGo/Subscription/Views/SubscriptionRestoreView.swift
#	DuckDuckGo/Subscription/Views/SubscriptionSettingsView.swift

afterxleep committed:
Fixes after merging main

afterxleep committed:
View onFirstAppear

afterxleep committed:
Return to settings from email view

afterxleep committed:
Dismiss Views from Email

afterxleep committed:
Cleanup

afterxleep committed:
Back to settings on managing email

afterxleep committed:
Pass navigationCoordinator to Subscription Views

afterxleep committed:
Add onFirstAppear and conditional URL loading

afterxleep committed:
Update navigation Coordinator to Allow both PopToSubscription Root & SettingsRoot

afterxleep committed:
Move Setup calls to onFirstAppear & onFirstDissapear

afterxleep committed:
Add links in settings

afterxleep committed:
Updated Methods to reflect Appear/Dissapear action

afterxleep committed:
Cleanup

afterxleep committed:
Remove local BSK

afterxleep committed:
Fix Initializer

afterxleep committed:
Update deeplinks

afterxleep committed:
Properly pass navigationCoordinator

afterxleep committed:
Hide title and content until data is loaded

afterxleep committed:
Extract manage device to other method

afterxleep committed:
Use memory cached subscription state to prevent UI glitches

afterxleep committed:
Unneded cache policy

federicocappelli commented on DuckDuckGo/Subscription/ViewModel/SubscriptionFlowViewModel.swift:
```diff
@@ -249,41 +250,23 @@ final class SubscriptionFlowViewModel: ObservableObject {
             self.resetState()
-            self.webViewModel.navigationCoordinator.navigateTo(url: self.purchaseURL )
         }
-        Pixel.fire(pixel: .privacyProOfferScreenImpression, debounce: 2)
```
side question, are you planning to remove the debounce feature from Pixel?

afterxleep replied:
We can remove from these pixels later, but no plans to remove from the class.  It may become useful later.

federicocappelli commented on DuckDuckGo/Subscription/Views/SubscriptionRestoreView.swift:
```diff
@@ -190,37 +181,39 @@ struct SubscriptionRestoreView: View {
-                        .foregroundColor(Color(designSystemColor: .textSecondary))
-                    getCellButton(buttonText: UserText.subscriptionActivateEmailButton,
-                                  action: {
-                        DailyPixel.fireDailyAndCount(pixel: .privacyProRestorePurchaseEmailStart)
```
My brain is in pain, too many changes XD, how confident are you that all pixels are still in place? without tests seems quite difficult to check

federicocappelli reviewed (commented):
Enormous PR, I would like @miasma13 to have a look too, LGTM for me, but the absence of tests makes me quite nervous about pixels and error management.

federicocappelli reviewed (approved):
LGTM for me

federicocappelli committed:
new pixel (#2661)

Task/Issue URL: https://app.asana.com/0/1205842942115003/1205469314884687/f

Description:
Added Missing privacyProRestorePurchaseClick pixel
